[ADS-276] Add samples to client-libs-java 1.0.4

iterativepianist opened the pull request.

iterativepianist committed:
[ADS-276] Add ConstantArgumentsBaseSample

iterativepianist committed:
[ADS-276] Add AllArgumentsSample

DonCziken commented on README.md:
```diff
@@ -0,0 +1,28 @@
+<h4> 1. The purpose of clisnt-libs-java-samples </h4>
```
typo `clisnt-libs-java-samples` -> `clients-libs-java-samples`

iterativepianist replied:
Thanks, fixed!

DonCziken commented on README.md:
```diff
@@ -0,0 +1,28 @@
+<h4> 1. The purpose of clisnt-libs-java-samples </h4>
+  
+All examples shown below fails, because samples does not contain appropriate adscore signature
```
lets add info that user need to fill this data with his own, ideally it should be a properties file ?

iterativepianist replied:
Added some info on how and to which example user needs to provide appropriate arguments.

For now, passing arguments to samples bases on command-line args:
e. g.
```
java -jar build/libs/client-libs-java-samples-1.0.4.jar all-args-sample {arg2} ... {argn...}
```
I believe is a bit simpler than using a properties file, but we can switch to use a properties file if necessary.

DonCziken commented on README.md:
```diff
@@ -0,0 +1,28 @@
+data.
+   
+<h4>2. Which classes are samples?</h4>
+All classes that implement Sample.java interface contain sample code.
```
we also should provide requirements to run sample, i.e. jdk XX

iterativepianist replied:
Added!

DonCziken commented on src/main/java/com/adscore/signature/sample/SignatureVerifierSampleMain.java:
```diff
@@ -0,0 +1,78 @@
+
+    if (args == null || args.length == 0 || samples.get(args[0]) == null) {
+      throw new SignatureVerifierSampleException(
+          "Not found sample for given sample name. Check README.md or provide correct sample name as a first argument.");
```
what does that mean ?:)

iterativepianist replied:
To execute any sample, the first java parameter should be the name of the sample.

```
java -jar build/libs/client-libs-java-samples-1.0.4.jar all-args-sample {arg2} ... {argn...} 
```

If user put:
```
java -jar build/libs/client-libs-java-samples-1.0.4.jar all-args-sample some-not-existing-sample-name {arg2} ... 
```
Then this exception is thrown.  (`samples.get(args[0]) == null` -> Sample does not exist in a sample map)

DonCziken commented on build.gradle:
```diff
@@ -0,0 +1,57 @@
+}
+
+dependencies {
+    compile 'com.adscore:adscore-client-libraries:1.0.4'
```
this also should be implementation (compile is deprecated)

iterativepianist replied:
Fixed!

iterativepianist replied:
Just figured out, what Gradle docs say:
`the implementation dependency configuration provides the same functionality as compile.` -> this is not true.

It makes difference when you build a fat jar (what we do here, to be allowed to just execute it and have all required dependencies inside). 

while using 'implementation' on a dependency, is not added to classpath while building a fat jar, therefore the sample is unexecutable :/

while using 'compile' here and then doing `gradle jar`, fat jar is packed properly, with com.github.seancfoley dependency inside it.
  
Base on this observation, I've reverted to using compile, because any way of building a fat jar by Gradle with 'implementation' dependency does not work. (checked with the latest version of Gradle, 6.8.3)

iterativepianist replied:
okay, finally we've left `implementation`, and the fat jar properly (by collecting dependencies on `configurations.runtimeClasspath` not `configurations.compile`

DonCziken commented on build.gradle:
```diff
@@ -0,0 +1,57 @@
+plugins {
+    id 'java-library'
```
java-library, signing, maven-publish and io.codearte.nexus-staging - are not needed as a plugins really, as this is not library but just a 'normal' project

iterativepianist replied:
True, only `java` and `spotless` plugins left.

DonCziken commented on build.gradle:
```diff
@@ -0,0 +1,57 @@
+dependencies {
+    compile 'com.adscore:adscore-client-libraries:1.0.4'
+
+    implementation 'com.github.seancfoley:ipaddress:5.2.1'
```
do we need this dep?

iterativepianist replied:
Well, as long as `adscore-client-libraries` is not a fat jar the answer is yes.

But I think we should make it a fat jar. I don`t know why it`s not it yet, but It does not sound good to force users of our library to provide required dependencies. At least it`s not a common practice for libraries.

I think making the lib as a fat jar, with full required dependencies (so only seancfoley:ipaddress for now) could be done as a separate commit for version 1.0.4.

DonCziken replied:
The library should not be a fat jar (https://stackoverflow.com/questions/50179525/for-official-maven-repositories-do-people-release-fat-jars). All required dependencies should be resolved by maven/gradle. I guess the issue here might be the fact that you are using id 'java-library'

iterativepianist replied:
Okay, the library should not be a fat jar, thanks for pointing to an answer that explains that.

But sample must be a fat jar, otherwise, the user is not able to execute it.

e.g 
If
`gradle jar` only packs sample classes,
then
`java -jar build/libs/client-libs-java-samples-1.0.4.jar base-sample` 

won`t work. end up with:
```
Exception in thread "main" java.lang.NoClassDefFoundError: com/adscore/signature/SignatureVerifier
```

(not using java-library plugin anymore).


I suggest not invest time into it. Works with compile + fatJar. When Gradle version 7.0.0 comes, we'll take care of switching from `compile` into `implementation`.

iterativepianist replied:
As written above, the problem with building a fat-jar is fixed.

The answer to the initial question: `do we need this dep?` is yes ;)

DonCziken replied:
well I've removed it from sample myself and wasn't needed ?:)

iterativepianist replied:
Oh, indeed it works. IDK how I was building it before... 
Thanks for checking that and correcting me!

DonCziken commented on README.md:
```diff
@@ -0,0 +1,42 @@
+<h4> 1. The purpose of clients-libs-java-samples </h4>
```
Lets also add initial section `What is this` with info such as: `this is a super simple CLI based sample showing how one could use Adscore API within his code.`

iterativepianist replied:
Added!

DonCziken replied:
Please also change MR desc :)

iterativepianist committed:
[ADS-276] Add multiple ip addresses to sample

Also, point to lib README and SignatureVerifier,
where the strategy of handling multiple ips is explained.

DonCziken commented on build.gradle:
```diff
@@ -0,0 +1,52 @@
+}
+
+group 'com.adscore'
+version '1.0.4' //version of samples match with the version of the client-libs-java project
```
I dont think we should 'bound' version of sample to version of library, the bound is represented in dependency.

iterativepianist replied:
Okay, I've set it to 1.0.0.

DonCziken replied:
:+1:
